drivers/ethernet/eth_nxp_enet.c

```diff
@@ -0,0 +1,1053 @@
+	ENET_Type *base;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	void (*generate_mac)(uint8_t *mac_addr);
+	const struct pinctrl_dev_config *pincfg;
+	enet_buffer_config_t buffer_config;
+	uint8_t phy_mode;
+	void (*irq_config_func)(void);
+	const struct device *phy_dev;
+	const struct device *mdio;
+#ifdef CONFIG_PTP_CLOCK_NXP_ENET
+	const struct device *ptp_clock;
+#endif
+};
+
+struct nxp_enet_mac_data {
+	struct net_if *iface;
+	uint8_t mac_addr[6];
+	enet_handle_t enet_handle;
+	struct k_sem tx_buf_sem;
+
+	K_KERNEL_STACK_MEMBER(rx_thread_stack, CONFIG_ETH_NXP_ENET_RX_THREAD_STACK_SIZE);
+
+	struct k_thread rx_thread;
+	struct k_sem rx_thread_sem;
+	struct k_mutex tx_frame_buf_mutex;
+	struct k_mutex rx_frame_buf_mutex;
+#ifdef CONFIG_PTP_CLOCK_NXP_ENET
+	struct k_sem ptp_ts_sem;
+	struct k_mutex *ptp_mutex;
+#endif
+	/* TODO: FIXME. This Ethernet frame sized buffer is used for
+	 * interfacing with MCUX. How it works is that hardware uses
+	 * DMA scatter buffers to receive a frame, and then public
+	 * MCUX call gathers them into this buffer (there's no other
+	 * public interface). All this happens only for this driver
+	 * to scatter this buffer again into Zephyr fragment buffers.
+	 * This is not efficient, but proper resolution of this issue
+	 * depends on introduction of zero-copy networking support
+	 * in Zephyr, and adding needed interface to MCUX (or
+	 * bypassing it and writing a more complex driver working
+	 * directly with hardware).
+	 *
+	 * Note that we do not copy FCS into this buffer thus the
+	 * size is 1514 bytes.
+	 */
+	uint8_t *tx_frame_buf; /* Max MTU + ethernet header */
+	uint8_t *rx_frame_buf; /* Max MTU + ethernet header */
+};
+
+/*
+ ********************
+ * Helper Functions *
+ ********************
+ */
+
+extern void nxp_enet_mdio_callback(const struct device *mdio_dev,
+				enum nxp_enet_callback_reason event);
+
+extern void nxp_enet_ptp_clock_callback(const struct device *dev,
+		enum nxp_enet_callback_reason event,
+		union nxp_enet_ptp_data *ptp_data);
+
+static inline struct net_if *get_iface(struct nxp_enet_mac_data *data, uint16_t vlan_tag)
+{
+#if defined(CONFIG_NET_VLAN)
+	struct net_if *iface;
+
+	iface = net_eth_get_vlan_iface(data->iface, vlan_tag);
+	if (!iface) {
+		return data->iface;
+	}
+
+	return iface;
+#else
+	ARG_UNUSED(vlan_tag);
+
+	return data->iface;
+#endif
+}
+
+#if defined(CONFIG_NET_IPV6)
+static void net_if_mcast_cb(struct net_if *iface,
+			    const struct net_addr *addr,
+			    bool is_joined)
+{
+	const struct device *dev = net_if_get_device(iface);
+	const struct nxp_enet_mac_config *config = dev->config;
+	struct net_eth_addr mac_addr;
+
+	if (addr->family != AF_INET6) {
+		return;
+	}
+
+	net_eth_ipv6_mcast_to_mac_addr(&addr->in6_addr, &mac_addr);
+
+	if (is_joined) {
+		ENET_AddMulticastGroup(config->base, mac_addr.addr);
+	} else {
+		ENET_LeaveMulticastGroup(config->base, mac_addr.addr);
+	}
+}
+#endif /* CONFIG_NET_IPV6 */
```

Comment:
See #65232 and add this to the PR

Reply:
Please be aware of https://github.com/zephyrproject-rtos/zephyr/pull/65550